File: researchflow-production-main/services/worker/src/api/routes/statistical_analysis.py
"""
Statistical Analysis API Route - Stage 7

FastAPI routes for StatisticalAnalysisAgent integration.
Handles statistical analysis requests from the orchestrator.

Linear Issues: ROS-XXX (Stage 7)
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
import logging
from datetime import datetime

# Import the statistical analysis agent
try:
    import sys
    from pathlib import Path
    # Add agents directory to path
    agents_dir = Path(__file__).parent.parent.parent.parent / "agents"
    sys.path.insert(0, str(agents_dir))
    
    from analysis.statistical_analysis_agent import (
        create_statistical_analysis_agent,
        StatisticalAnalysisAgent,
    )
    from analysis.statistical_types import (
        StudyData,
        StatisticalResult,
    )
    AGENT_AVAILABLE = True
except ImportError as e:
    AGENT_AVAILABLE = False
    logging.warning(f"StatisticalAnalysisAgent not available: {e}")

logger = logging.getLogger(__name__)
router = APIRouter()


# =============================================================================
# Request/Response Models
# =============================================================================

class StudyDataRequest(BaseModel):
    """Study data for statistical analysis"""
    groups: Optional[List[str]] = Field(None, description="Group labels")
    outcomes: Dict[str, List[float]] = Field(..., description="Outcome variables")
    covariates: Optional[Dict[str, List[Any]]] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class AnalysisOptions(BaseModel):
    """Optional analysis configuration"""
    test_type: Optional[str] = None
    confidence_level: float = 0.95
    alpha: float = 0.05
    calculate_effect_size: bool = True
    check_assumptions: bool = True
    generate_visualizations: bool = True


class StatisticalAnalysisRequest(BaseModel):
    """Complete statistical analysis request"""
    study_data: StudyDataRequest
    options: Optional[AnalysisOptions] = None


class StatisticalAnalysisResponse(BaseModel):
    """Statistical analysis response"""
    success: bool
    result: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    duration_ms: int
    agent_iterations: int
    quality_score: float


# =============================================================================
# Routes
# =============================================================================

@router.post("/api/analysis/statistical")
async def run_statistical_analysis(request: StatisticalAnalysisRequest):
    """
    Execute statistical analysis using StatisticalAnalysisAgent.
    
    This is the primary endpoint for Stage 7 statistical analysis.
    Called by the orchestrator service.
    
    Args:
        request: Statistical analysis request with study data and options
    
    Returns:
        Statistical analysis result with descriptive stats, inferential tests,
        effect sizes, assumption checks, and APA-formatted tables
    """
    start_time = datetime.utcnow()
    
    if not AGENT_AVAILABLE:
        raise HTTPException(
            status_code=503,
            detail="StatisticalAnalysisAgent not available. Check dependencies."
        )
    
    try:
        # Create agent instance
        agent = create_statistical_analysis_agent()
        
        # Convert request to StudyData
        study_data = StudyData(
            groups=request.study_data.groups,
            outcomes=request.study_data.outcomes,
            covariates=request.study_data.covariates,
            metadata=request.study_data.metadata,
        )
        
        logger.info(f"Running statistical analysis for research: {study_data.metadata.get('research_id')}")
        
        # Execute analysis
        result = await agent.execute(study_data)
        
        # Calculate duration
        duration_ms = int((datetime.utcnow() - start_time).total_seconds() * 1000)
        
        logger.info(f"Statistical analysis completed in {duration_ms}ms")
        
        return StatisticalAnalysisResponse(
            success=True,
            result=result.to_dict() if result else None,
            error=None,
            duration_ms=duration_ms,
            agent_iterations=0,  # TODO: Track from agent
            quality_score=0.0,   # TODO: Get from agent result
        )
    
    except Exception as e:
        logger.error(f"Statistical analysis failed: {e}")
        duration_ms = int((datetime.utcnow() - start_time).total_seconds() * 1000)
        
        return StatisticalAnalysisResponse(
            success=False,
            result=None,
            error=str(e),
            duration_ms=duration_ms,
            agent_iterations=0,
            quality_score=0.0,
        )


@router.get("/api/analysis/statistical/capabilities")
async def get_capabilities():
    """
    Get statistical analysis capabilities.
    
    Returns:
        Information about available tests, assumptions, and features
    """
    if not AGENT_AVAILABLE:
        return {
            "available": False,
            "error": "StatisticalAnalysisAgent not loaded",
        }
    
    return {
        "available": True,
        "agent": "StatisticalAnalysisAgent",
        "version": "1.0.0",
        "capabilities": {
            "hypothesis_tests": [
                "t_test_independent",
                "t_test_paired",
                "mann_whitney",
                "wilcoxon",
                "anova_oneway",
                "kruskal_wallis",
                "chi_square",
            ],
            "effect_sizes": [
                "cohens_d",
                "hedges_g",
                "eta_squared",
            ],
            "assumption_checks": [
                "normality",
                "homogeneity",
                "independence",
            ],
            "outputs": [
                "descriptive_statistics",
                "hypothesis_test_results",
                "effect_sizes",
                "confidence_intervals",
                "apa_formatted_tables",
                "visualization_specifications",
                "remediation_suggestions",
            ],
        },
        "features": {
            "multiple_groups": True,
            "assumption_remediation": True,
            "visualization_specs": True,
            "apa_formatting": True,
            "quality_gates": True,
        },
    }


@router.get("/api/analysis/statistical/health")
async def health_check():
    """
    Health check for statistical analysis service.
    
    Returns:
        Service status and dependencies
    """
    status = {
        "service": "statistical-analysis",
        "status": "healthy" if AGENT_AVAILABLE else "unavailable",
        "agent_loaded": AGENT_AVAILABLE,
        "timestamp": datetime.utcnow().isoformat(),
    }
    
    # Check dependencies
    dependencies = {}
    try:
        import scipy
        dependencies["scipy"] = scipy.__version__
    except ImportError:
        dependencies["scipy"] = "not installed"
    
    try:
        import statsmodels
        dependencies["statsmodels"] = statsmodels.__version__
    except ImportError:
        dependencies["statsmodels"] = "not installed"
    
    try:
        import pandas
        dependencies["pandas"] = pandas.__version__
    except ImportError:
        dependencies["pandas"] = "not installed"
    
    try:
        import numpy
        dependencies["numpy"] = numpy.__version__
    except ImportError:
        dependencies["numpy"] = "not installed"
    
    status["dependencies"] = dependencies
    
    return status

File: researchflow-production-main/services/orchestrator/src/routes/statistical-analysis.ts
/**
 * Statistical Analysis Route - Stage 7
 * 
 * Handles Stage 7 (Statistical Analysis) execution including:
 * - Descriptive statistics
 * - Hypothesis testing (t-tests, ANOVA, chi-square, non-parametric)
 * - Effect size calculations
 * - Assumption checking with remediation
 * - APA-style reporting
 * 
 * Routes to StatisticalAnalysisAgent in worker service.
 */

import { Router, Request, Response } from 'express';
import * as z from 'zod';

import { config } from '../config/env';
import { asyncHandler } from '../middleware/asyncHandler';
import { requirePermission } from '../middleware/rbac';
import { logAction } from '../services/audit-service';
import { createLogger } from '../utils/logger';

const router = Router();
const logger = createLogger('statistical-analysis-route');

// Worker service URL
const WORKER_URL = config.workerUrl;

// =============================================================================
// Request Schemas
// =============================================================================

/**
 * Study data schema matching Python StudyData type
 */
const StudyDataSchema = z.object({
  groups: z.array(z.string()).optional().describe('Group labels for each observation'),
  outcomes: z.record(z.array(z.number())).describe('Outcome variables as dict of variable_name -> values'),
  covariates: z.record(z.union([z.array(z.number()), z.array(z.string())])).optional().describe('Optional covariates'),
  metadata: z.record(z.unknown()).default({}).describe('Study metadata'),
});

/**
 * Statistical analysis request schema
 */
const StatisticalAnalysisRequestSchema = z.object({
  study_data: StudyDataSchema,
  research_id: z.string().optional(),
  analysis_id: z.string().optional(),
  options: z.object({
    test_type: z.enum([
      't_test_independent',
      't_test_paired',
      'mann_whitney',
      'wilcoxon',
      'anova_oneway',
      'kruskal_wallis',
      'chi_square',
    ]).optional().describe('Specific test type to use (auto-detected if not provided)'),
    confidence_level: z.number().min(0.5).max(0.99).default(0.95),
    alpha: z.number().min(0.001).max(0.1).default(0.05),
    calculate_effect_size: z.boolean().default(true),
    check_assumptions: z.boolean().default(true),
    generate_visualizations: z.boolean().default(true),
  }).optional(),
});

// =============================================================================
// Routes
// =============================================================================

/**
 * POST /api/research/:id/stage/7/execute
 * 
 * Execute statistical analysis for Stage 7 of research workflow.
 * 
 * This is the primary endpoint for running StatisticalAnalysisAgent.
 */
router.post(
  '/research/:id/stage/7/execute',
  requirePermission('ANALYZE'),
  asyncHandler(async (req: Request, res: Response) => {
    const startTime = Date.now();
    const researchId = req.params.id;
    const requestId = `stat_analysis_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;

    logger.info('Statistical analysis requested', { researchId, requestId });

    // Validate request
    const parseResult = StatisticalAnalysisRequestSchema.safeParse(req.body);
    if (!parseResult.success) {
      logger.warn('Invalid statistical analysis request', { 
        researchId, 
        errors: parseResult.error.issues 
      });

      return res.status(400).json({
        error: 'INVALID_REQUEST',
        message: 'Invalid statistical analysis request',
        details: parseResult.error.issues,
        request_id: requestId,
      });
    }

    const { study_data, analysis_id, options } = parseResult.data;

    // Enrich metadata with research context
    study_data.metadata = {
      ...study_data.metadata,
      research_id: researchId,
      analysis_id: analysis_id || requestId,
      requested_by: req.user?.id || 'anonymous',
      requested_at: new Date().toISOString(),
    };

    try {
      // Call worker's statistical analysis endpoint
      logger.info('Calling worker statistical analysis service', { 
        researchId, 
        workerUrl: WORKER_URL 
      });

      const workerResponse = await fetch(`${WORKER_URL}/api/analysis/statistical`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Request-ID': requestId,
          'X-Research-ID': researchId,
        },
        body: JSON.stringify({
          study_data,
          options,
        }),
        signal: AbortSignal.timeout(300000), // 5 minute timeout
      });

      if (!workerResponse.ok) {
        const errorText = await workerResponse.text();
        logger.error('Worker statistical analysis failed', {
          researchId,
          status: workerResponse.status,
          error: errorText,
        });

        return res.status(workerResponse.status).json({
          error: 'ANALYSIS_FAILED',
          message: `Worker returned ${workerResponse.status}`,
          details: errorText,
          request_id: requestId,
        });
      }

      const analysisResult = await workerResponse.json();

      // Log successful analysis
      await logAction({
        userId: req.user?.id || 'anonymous',
        action: 'STATISTICAL_ANALYSIS',
        resourceType: 'research',
        resourceId: researchId,
        metadata: {
          analysis_id: analysis_id || requestId,
          stage: 7,
          sample_size: study_data.groups?.length || 0,
          outcome_variables: Object.keys(study_data.outcomes || {}),
          test_performed: analysisResult.inferential?.test_type || 'unknown',
          quality_score: analysisResult.quality_score || null,
          duration_ms: Date.now() - startTime,
        },
      });

      logger.info('Statistical analysis completed successfully', {
        researchId,
        requestId,
        durationMs: Date.now() - startTime,
        testType: analysisResult.inferential?.test_type,
      });

      return res.json({
        request_id: requestId,
        research_id: researchId,
        status: 'completed',
        result: analysisResult,
        duration_ms: Date.now() - startTime,
      });

    } catch (error) {
      logger.error('Statistical analysis error', {
        researchId,
        requestId,
        error: error instanceof Error ? error.message : 'Unknown error',
      });

      return res.status(503).json({
        error: 'SERVICE_UNAVAILABLE',
        message: error instanceof Error ? error.message : 'Worker service unavailable',
        request_id: requestId,
      });
    }
  })
);

/**
 * POST /api/analysis/statistical/validate
 * 
 * Validate study data before running analysis.
 * Checks data quality, sample sizes, missing values, etc.
 */
router.post(
  '/analysis/statistical/validate',
  requirePermission('ANALYZE'),
  asyncHandler(async (req: Request, res: Response) => {
    const requestId = `validate_${Date.now()}`;

    // Validate request schema
    const parseResult = StudyDataSchema.safeParse(req.body);
    if (!parseResult.success) {
      return res.status(400).json({
        error: 'INVALID_DATA',
        message: 'Invalid study data format',
        details: parseResult.error.issues,
        request_id: requestId,
      });
    }

    const study_data = parseResult.data;

    // Basic validation checks
    const warnings: string[] = [];
    const errors: string[] = [];

    // Check sample size
    const sampleSize = study_data.groups?.length || 
                       (Object.values(study_data.outcomes)[0]?.length || 0);
    
    if (sampleSize < 3) {
      errors.push('Sample size too small (n < 3)');
    } else if (sampleSize < 10) {
      warnings.push('Small sample size (n < 10) - some tests may be unreliable');
    }

    // Check for missing values
    for (const [varName, values] of Object.entries(study_data.outcomes || {})) {
      const nullCount = values.filter(v => v === null || v === undefined).length;
      if (nullCount > 0) {
        warnings.push(`${varName}: ${nullCount} missing values (${(nullCount/values.length*100).toFixed(1)}%)`);
      }
    }

    // Check group balance (if groups provided)
    if (study_data.groups && study_data.groups.length > 0) {
      const groupCounts = study_data.groups.reduce((acc, g) => {
        acc[g] = (acc[g] || 0) + 1;
        return acc;
      }, {} as Record<string, number>);

      const groupSizes = Object.values(groupCounts);
      const maxSize = Math.max(...groupSizes);
      const minSize = Math.min(...groupSizes);

      if (maxSize / minSize > 2) {
        warnings.push('Groups are unbalanced (size ratio > 2:1)');
      }
    }

    const isValid = errors.length === 0;

    return res.json({
      request_id: requestId,
      valid: isValid,
      warnings,
      errors,
      metadata: {
        sample_size: sampleSize,
        outcome_variables: Object.keys(study_data.outcomes || {}).length,
        has_groups: Boolean(study_data.groups),
        has_covariates: Boolean(study_data.covariates),
      },
    });
  })
);

/**
 * GET /api/analysis/statistical/tests
 * 
 * List available statistical tests with descriptions.
 */
router.get('/analysis/statistical/tests', (_req: Request, res: Response) => {
  res.json({
    tests: [
      {
        type: 't_test_independent',
        name: 'Independent t-test',
        description: 'Compare means of two independent groups',
        assumptions: ['normality', 'homogeneity'],
        sample_size_min: 3,
        groups: 2,
      },
      {
        type: 't_test_paired',
        name: 'Paired t-test',
        description: 'Compare means of two paired groups',
        assumptions: ['normality'],
        sample_size_min: 3,
        groups: 2,
      },
      {
        type: 'mann_whitney',
        name: 'Mann-Whitney U test',
        description: 'Non-parametric alternative to independent t-test',
        assumptions: ['independence'],
        sample_size_min: 3,
        groups: 2,
      },
      {
        type: 'wilcoxon',
        name: 'Wilcoxon signed-rank test',
        description: 'Non-parametric alternative to paired t-test',
        assumptions: ['independence'],
        sample_size_min: 3,
        groups: 2,
      },
      {
        type: 'anova_oneway',
        name: 'One-way ANOVA',
        description: 'Compare means of three or more independent groups',
        assumptions: ['normality', 'homogeneity'],
        sample_size_min: 3,
        groups: '3+',
      },
      {
        type: 'kruskal_wallis',
        name: 'Kruskal-Wallis H test',
        description: 'Non-parametric alternative to one-way ANOVA',
        assumptions: ['independence'],
        sample_size_min: 3,
        groups: '3+',
      },
      {
        type: 'chi_square',
        name: 'Chi-square test of independence',
        description: 'Test association between categorical variables',
        assumptions: ['expected_counts >= 5'],
        sample_size_min: 5,
        groups: '2+',
      },
    ],
  });
});

/**
 * GET /api/analysis/statistical/health
 * 
 * Health check for statistical analysis service.
 */
router.get('/analysis/statistical/health', async (_req: Request, res: Response) => {
  let workerStatus = 'unknown';

  try {
    const workerResponse = await fetch(`${WORKER_URL}/health`, {
      method: 'GET',
      signal: AbortSignal.timeout(5000),
    });
    workerStatus = workerResponse.ok ? 'healthy' : 'unhealthy';
  } catch {
    workerStatus = 'unreachable';
  }

  res.json({
    status: 'healthy',
    service: 'statistical-analysis',
    stage: 7,
    worker_url: WORKER_URL,
    worker_status: workerStatus,
    timestamp: new Date().toISOString(),
  });
});

export default router;
